docs/specs/saml.md

```diff
@@ -0,0 +1,697 @@
+- In step [5], the SP should memorize the authentication state. For example, by creating a session using cookies.
+
+### <a id="1_2"></a> IdP Initiated
+
+Ref: [SAMLProfile](https://groups.oasis-open.org/higherlogic/ws/public/download/35389/sstc-saml-profiles-errata-2.0-wd-06-diff.pdf) Section 4.1
+
+```mermaid
+sequenceDiagram
+    participant ua as User Agent (Alice)
+    participant sp as SP (Web App) <br/> https://example.com
+    participant ag as IdP (Authgear) <br/> https://example.authgear.cloud
+    ua->>+sp: [1] Visit https://example.com/
+    sp->>-ua: [2] Redirect to https://example.authgear.cloud/saml2/login/ENTITY_ID
+    ua->>+ag: [3] GET https://example.authgear.cloud/saml2/login/ENTITY_ID
+    Note right of ag: Alice enters user credentials
+    ag->>-ua: [4] Redirect to https://example.com/acs with <Response>
+    ua->>+sp: [5] Send <Response> to https://example.com/acs
+    Note right of sp: Authenticated as Alice
+    sp->>-ua: [6] Redirect to https://example.com/
+
+```
+
+- The login endpoint `https://example.authgear.cloud/saml2/login/ENTITY_ID` is from the metadata. Different applications will use a different endpoint.
+- Same as SP Initiated Web Browser SSO, Idp Session is created in step [4]. `SessionIndex` in the `<Response>` of step [5] is the ID of the IdP Session.
+- Same as SP Initiated Web Browser SSO, SP should memorize the authentication state in step [5] .
+- The Assertion Consumer Service (ACS) URL `https://example.com/acs` MUST be configured in the application config. Config name `acs_url`.
+- `Audience` of the MUST be configured in the application config. Config name `audience`.
+- The URL the user finally reached in step [6] is up to the implementaion of the SP. In the above example, it backs to the home page `https://example.com/`.
+
+### <a id="1_3"></a> Configs
+
+Here is an example of configurations about Web Browser SSO:
+
+```yaml
+saml:
+  service_providers:
+    - id: ENTITY_ID
+      audience: https://example.com
+      destination: https://example.com/acs
+      acs_url: https://example.com/acs
+```
+
+- `saml.service_providers`: Required. A list of objects containing the configurations of a Service Provider. The objects in the list contains the following fields:
+  - `id`: Required. A string which is the entity ID of the service provider.
+  - `audience`: Optional in SP Initiated flow. Required when using the IdP Initiated flow. `Audience` of the `<Response>` generated at the end of Browser SSO flow. In SP Initiated flow, if this config is set, the value configured will be used instead of `Issuer` of the `<AuthnRequest>`.
+  - `destination`: Optional in SP Initiated flow. Required when using the IdP Initiated flow. `Destination` of the `<Response>` generated at the end of Browser SSO flow. In SP Initiated flow, if this config is set, the value configured will be used instead of `AssertionConsumerServiceURL` of the `<AuthnRequest>`.
```

Comment:
I made a mistake here. `AssertionConsumerServiceURL` must always be configured, no matter in SP or IdP initiated flow. And, we will always validate the `AssertionConsumerServiceURL` in request matches the config. So there is no problem of overriding (Because values are always same)